### PR TITLE
Move '@encrypted' into '@graph'

### PR DESCRIPTION
**Pull request by JLoveUOA**
Based on feedback from the RO-Crate group regarding compliance and usability the following changes have been made:

- content that was previously listed in a new section of JSON-LD (`@encrypted`) is now within `@graph`
-  Messages containing encrypted data have been defined as a  `["SendAction", "EncryptedGraphMessage"]`
- public keys are now associated with `recipients`. recipients may be any context object that are listed via the property `recipients` on an `EncryptedGrpahMessage` or `EncryptedContextEntity`.
- Public Keys on Recipients are collected during encryption and used to encrypt `EncryptedContextEntities` into `EncryptedGraphMessage`s. they are stored on the custom property `pubkey_fingerprints`
- A custom `Keyholder` type has been added to handle cases where pubkey fingerprints are held by an entity that is not a `person` or an `organisation` with some utilities for handling key identifiers and retrieval from key-servers. 
- The URI for a keyholder SHOULD be <keyserver><request-parameters><pubkeyfingerprint> in a form that uniquely retrieves one public key. However if a keyserver is not available a 40 byte type4 pubkeyfingerprint is acceptable.
- Tests have been updated to reflect these new types.
- new errors for missing keys and missing recipients have been added. These should be handled differently based on use case (although intended behavior is that an encrypted context entity without any recipients or keys will never be written and should throw and error)

TODO:

- add a profile to conformTo
- Signatures

### CODE REVIEW

File: .vscode/settings.json
Comment (ChrisKSeal) [inline .vscode/settings.json:1]: Probably should be in a .gitignore since this is a local VS-Code settings file :)

File: poetry.lock
Comment (ChrisKSeal) [inline poetry.lock:1]: Nothing to review here

File: rocrate/metadata.py
Comment (ChrisKSeal) [inline rocrate/metadata.py:1]: Should we include UoA in the list of copyright holders?

File: rocrate/metadata.py [hunk 1]
```diff
@@ -1,8 +1,9 @@
...
-# Copyright 2020-2023 Vlaams Instituut voor Biotechnologie (VIB), BE
-# Copyright 2020-2023 Barcelona Supercomputing Center (BSC), ES
-# Copyright 2020-2023 Center for Advanced Studies, Research and Development in Sardinia (CRS4), IT
-# Copyright 2022-2023 École Polytechnique Fédérale de Lausanne, CH
+# Copyright 2019-2024 The University of Manchester, UK
+# Copyright 2020-2024 Vlaams Instituut voor Biotechnologie (VIB), BE
+# Copyright 2020-2024 Barcelona Supercomputing Center (BSC), ES
+# Copyright 2020-2024 Center for Advanced Studies, Research and Development in Sardinia (CRS4), IT
+# Copyright 2022-2024 École Polytechnique Fédérale de Lausanne, CH
+# Copyright 2024 Data Centre, SciLifeLab, SE
```
Comment (ChrisKSeal) [inline rocrate/metadata.py:6]: Should we include UoA in this list?

File: rocrate/model/keyholder.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                    if keyserver:
+                        identifier = f"{keyserver}{HPK_STUB}{pubkey_fingerprint.key}"
+                    else:
+                        identifier = pubkey_fingerprint.key
+                else:
+                    raise ValueError(f"No valid identifier combination supplied for keyholder")
+            if pubkey_fingerprint:
+                names, emails =  zip(*[split_uid(uid) for uid in pubkey_fingerprint.uids])
+                properties["pubkey_fingerprints"] = pubkey_fingerprint.key
+                properties["email"] = [email for email in emails if email != ""]
```
Comment (ChrisKSeal) [inline rocrate/model/keyholder.py]: There might be a case where, if one person doesn't have an email address, there is a shift in the list and we assign the wrong email to subsequent people. Have you considered this? If not, can I suggest we add an 'or "No Email"' or similar to the end of your list comprehension?

File: rocrate/encryption_utils.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+class MissingMemberException(Exception):
+    "Raised if sigining to a set of members and one is missing a key"
+    def __init__(self, message):            
+        super().__init__(message)
+    def __str__(self):
+        return f'At least one target lacks a valid key, or cannot be found in the graph'
+
+def combine_recipient_keys(target_entity: Entity, allow_missing: Optional[bool]=False) -> list[str]:
+    """Retrun the complete set of all keys found on this entity and it's recipients
```
Comment (ChrisKSeal) [inline rocrate/encryption_utils.py]: Typo in Return

File: rocrate/model/metadata.py [hunk 4]
```diff
@@ -127,23 +130,28 @@ def __encrypt_fields(self, encrypted_fields: Dict[List[str],List[Dict[str,str]]]
...
-                    uids =current_keys.key_map[fingerprint].get("uids") or [],
-                    key=fingerprint)
-                    for fingerprint in fingerprints],
-                encrypted_field._as_text(),
-                method="https://doi.org/10.17487/RFC4880",
+                crate= self.crate,
+                identifier=f"Encrypted_Message{"_".join(fingerprints)}",
+                encrypted_graph=encrypted_field._as_text(),
+                properties={
+                    "deliveryMethod":"https://doi.org/10.17487/RFC4880",
```
Comment (ChrisKSeal) [inline rocrate/model/metadata.py:152]: Nice :)

File: test/test_encryption.py [hunk 5]
```diff
@@ -47,169 +50,200 @@ def test_gpg_key_2(test_gpg_object: GPG, test_passphrase:str):
...
+
+    )
+
+
+@pytest.fixture
+def test_pubkey_nosecret(test_gpg_object: GPG, test_gpg_key) -> PubkeyObject:
+    return PubkeyObject(
+        method="1",
+        key="643BE0B0159AEFC2B9428D578533F9C66A9E7628",
+        uids=[""]
```
Comment (ChrisKSeal) [inline test/test_encryption.py]: Should we have something in the uids field to test the uids extraction function?

File: test/test_encryption.py
Comment (ChrisKSeal) [inline test/test_encryption.py:1]: Are we confident that we have covered the codebase sufficiently? We do some wrangling of ids based on GPG - have we tested these. It may pay to mock some responses from the keyserver to make sure that we handle returns or non-returns properly? Might pay to run coverage over the tests to see what percentage of the code base is run.

**Review by ChrisKSeal — APPROVED** [review]
All good and ready to merge. I've added a few comments. There is one case where I think it'd be good to verify that we're not going to get emails out of sync from users, where they are in separate lists, which may need checking. I also wonder if we have sufficient tests in place, particularly around error handling if we get an empty return from the keyserver. Happy to chat about these.
